### DISCUSSION TITLE
Running strongSwan inside a Network Namespace

### DISCUSSION THREAD

**Discussion by gr4yg00s3** [strongswan/strongswan#1137] [General]
I have a couple of questions regarding the guide below:

```
Luckily, the ip netns exec command provides a helpful feature: Every file found in /etc/netns/<name>/ for a given netns is bind-mounted over its corresponding counterpart in /etc/ (so it has to exist there). This can be used to provide different config files for each instance but may also be used to redirect the so called piddir where the [charon](https://docs.strongswan.org/docs/5.9/daemons/charon.html) daemon creates its PID file and UNIX sockets. The default is to use /var/run which would conflict if multiple instances would use it.

In order to fix this, make sure strongSwan is [configured](https://docs.strongswan.org/docs/5.9/install/autoconf.html) with --sysconfdir=/etc and e.g. --with-piddir=/etc/ipsec.d/run. Then after [building and installing strongSwan](https://docs.strongswan.org/docs/5.9/install/install.html#_building_strongswan) the piddirs can be created as follows:

# mkdir -p /etc/ipsec.d/run
# mkdir -p /etc/netns/<netns name 1>/ipsec.d/run
# mkdir -p /etc/netns/<netns name 2>/ipsec.d/run
```

1. When charon is started from a particular net ns are we expected to see the pid files and sockets in ```/etc/netns/<netns name 2>/ipsec.d/run```? If yes, I do not see it. Any pointers as to what I can try?
2. In order to provide different config files for each instance are we supposed to create the config files in ```/etc/netns/<name>/swanctl/conf.d```? It says above that the file also has to exist in ```/etc/``` does that mean there has to be a copy created there as well?

**Comment by tobiasbrunner** [comment]
> 1.  When charon is started from a particular net ns are we expected to see the pid files and sockets in /etc/netns/<netns name 2>/ipsec.d/run?

While the daemon is running you should see the pid file. Check if it actually is running (and in that particular NS). If you changed the configure options after building once, make sure to run `make clean` before rebuilding and installing again. The UNIX sockets crated by plugins should be in the same directory unless the paths/URIs were changed via strongswan.conf.

> 2. In order to provide different config files for each instance are we supposed to create the config files in /etc/netns/<name>/swanctl/conf.d? It says above that the file also has to exist in /etc/ does that mean there has to be a copy created there as well?

All files and directories in `/etc/netns/<name>/` have to exist in `/etc` as they are simply bind-mounted over their counterparts. However, files in sub-directories don't have to exist, as the complete top-level directory is "replaced". So in your example you only need to create `/etc/swanctl` to make `/etc/netns/<name>/swanctl/conf.d` and all files it contains available in the NS.

**Reply by gr4yg00s3** [reply]
`make clean` worked 👍

Thank you, this was helpful!
